File: Python/4.py
# 4. Median of Two Sorted Arrays

# There are two sorted arrays nums1 and nums2 of size m and n respectively.

# Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).

# You may assume nums1 and nums2 cannot be both empty.

# Example 1:

# nums1 = [1, 3]
# nums2 = [2]

# The median is 2.0
# Example 2:

# nums1 = [1, 2]
# nums2 = [3, 4]

# The median is (2 + 3)/2 = 2.5
from typing import List


class Solution:
    def find_k_th(self, nums1, nums2, k):
        """
        find k-th value from nums1 and nums2
        """
        m = len(nums1)
        n = len(nums2)
        # expect n is equal or smaller than m
        if m > n:
            return self.find_k_th(nums2, nums1, k)
        if m == 0:
            return nums2[k - 1]
        if k == 1:
            return min(nums1[0], nums2[0])
        p1 = min(int(k / 2), m)
        p2 = k - p1
        if nums1[p1 - 1] <= nums2[p2 - 1]:
            return self.find_k_th(nums1[p1:], nums2, p2)
        else:
            return self.find_k_th(nums1, nums2[p2:], p1)
        # else:
        # return nums1[]

    def findMedianSortedArrays(self, nums1: List[int],
                               nums2: List[int]) -> float:
        """
        if the length is odd  ,median is element at (length / 2 + 1)
        if the length is even ,median is the median of (length / 2) and (length / 2 + 1)
        """
        length = len(nums1) + len(nums2)
        mid = length >> 1
        mid_1 = self.find_k_th(nums1, nums2, mid + 1)
        if length % 2 != 0:
            return mid_1
        return (mid_1 + self.find_k_th(nums1, nums2, mid)) / 2


if __name__ == '__main__':
    from util import Test
    s = Solution()
    t = Test(s.findMedianSortedArrays)
    t.equal(2, [1, 2], [2])
    t.equal(1, [1], [2])

File: Golang/4/4.go
package main

import (
	"fmt"
)

func min(x, y int) int {
	if x > y {
		return y
	}
	return x
}

func findKth(nums1, nums2 []int, k int) int {
	m := len(nums1)
	n := len(nums2)
	// we assume that len(nums2) is greater than len(nums1)
	if m > n {
		return findKth(nums2, nums1, k)
	}
	if m == 0 {
		return nums2[k-1]
	}
	if k == 1 {
		return min(nums2[0], nums1[0])
	}
	p1 := min(k>>1, m)
	p2 := k - p1
	if nums1[p1-1] == nums2[p2-1] {
		return nums1[p1-1]
	} else if nums1[p1-1] < nums2[p2-1] {
		return findKth(nums1[p1:], nums2, p2)
	}
	return findKth(nums1, nums2[p2:], p1)
}

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	m := len(nums1)
	n := len(nums2)
	// even number
	if (m+n)%2 == 0 {
		return float64(findKth(nums1, nums2, ((m+n)>>1)+1)+findKth(nums1, nums2, (m+n)>>1)) * 0.5
	}
	return float64(findKth(nums1, nums2, (m+n)>>1+1))
}

func main() {
	fmt.Printf("median of [1,3] and [2] is %v \n", findMedianSortedArrays([]int{1, 3}, []int{2}))
	fmt.Printf("median of [1,2] and [3,4] is %v \n", findMedianSortedArrays([]int{1, 2}, []int{3, 4}))
	fmt.Printf("median of [] and [1] is %v \n", findMedianSortedArrays([]int{}, []int{1}))
}
